includes/class-wc-order-item-coupon.php

```diff
@@ -0,0 +1,187 @@
+			$offset = 'discount';
+		} elseif ( 'discount_amount_tax' === $offset ) {
+			$offset = 'discount_tax';
+		}
+		return parent::offsetGet( $offset );
+	}
+
+	/**
+	 * offsetSet for ArrayAccess/Backwards compatibility.
+	 * @deprecated Add deprecation notices in future release.
+	 * @param string $offset
+	 * @param mixed $value
+	 */
+	public function offsetSet( $offset, $value ) {
+		if ( 'discount_amount' === $offset ) {
+			$offset = 'discount';
+		} elseif ( 'discount_amount_tax' === $offset ) {
+			$offset = 'discount_tax';
+		}
+		parent::offsetSet( $offset, $value );
+	}
+
+	/**
+	 * offsetExists for ArrayAccess
+	 * @param string $offset
+	 * @return bool
+	 */
+	public function offsetExists( $offset ) {
+		if ( in_array( $offset, array( 'discount_amount', 'discount_amount_tax' ) ) ) {
+			return true;
+		}
+		return parent::offsetExists( $offset );
+	}
+
+	/**
+	 * Read/populate data properties specific to this order item.
+	 */
+	public function read( $id ) {
+		parent::read( $id );
+		if ( $this->get_id() ) {
+			$this->set_discount( get_metadata( 'order_item', $this->get_id(), 'discount_amount', true ) );
+			$this->set_discount_tax( get_metadata( 'order_item', $this->get_id(), 'discount_amount_tax', true ) );
+		}
+	}
+
+	/**
+	 * Save properties specific to this order item.
+	 * @return int Item ID
+	 */
+	public function save() {
+		parent::save();
+		if ( $this->get_id() ) {
+			wc_update_order_item_meta( $this->get_id(), 'discount_amount', $this->get_discount() );
+			wc_update_order_item_meta( $this->get_id(), 'discount_amount_tax', $this->get_discount_tax() );
+		}
+
+		return $this->get_id();
+	}
+
+	/**
+	 * Internal meta keys we don't want exposed as part of meta_data.
+	 * @return array()
+	 */
+	protected function get_internal_meta_keys() {
+		return array( 'discount_amount', 'discount_amount_tax' );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Set order item name.
+	 * @param string $value
+	 */
+	public function set_name( $value ) {
+		$this->set_code( $value );
+	}
+
+	/**
+	 * Set code.
+	 * @param string $value
+	 */
+	public function set_code( $value ) {
+		$this->_data['code'] = wc_clean( $value );
+	}
+
+	/**
+	 * Set discount amount.
+	 * @param string $value
+	 */
+	public function set_discount( $value ) {
+		$this->_data['discount'] =  wc_format_decimal( $value );
+	}
+
+	/**
+	 * Set discounted tax amount.
+	 * @param string $value
+	 */
+	public function set_discount_tax( $value ) {
+		$this->_data['discount_tax'] = wc_format_decimal( $value );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Getters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Get order item type.
+	 * @return string
+	 */
+	public function get_type() {
+		return 'coupon';
```

Comment:
`set_type` has never been a supported function with these items has it? In which case we could just remove that and have it completely read only to prevent what you pointed out.